flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/OneInputStreamTaskTest.java

```diff
@@ -597,6 +603,54 @@ public void testQuiesceTimerServiceAfterOpClose() throws Exception {
 		timeService.shutdownService();
 	}
 
+	@Test
+	public void testHandlingEndOfInput() throws Exception {
+		final OneInputStreamTaskTestHarness<String, String> testHarness = new OneInputStreamTaskTestHarness<>(
+			OneInputStreamTask::new,
+			2, 2,
+			BasicTypeInfo.STRING_TYPE_INFO, BasicTypeInfo.STRING_TYPE_INFO);
+
+		testHarness
+			.setupOperatorChain(new OperatorID(), new TestBoundedOneInputOperator("Operator0"))
+			.chain(
+				new OperatorID(),
+				new TestBoundedOneInputOperator("Operator1"),
+				BasicTypeInfo.STRING_TYPE_INFO.createSerializer(new ExecutionConfig()))
+			.finish();
+
+		ConcurrentLinkedQueue<Object> expectedOutput = new ConcurrentLinkedQueue<>();
+
+		testHarness.invoke();
+		testHarness.waitForTaskRunning();
+
+		TestBoundedOneInputOperator headOperator = (TestBoundedOneInputOperator) testHarness.getTask().headOperator;
+
+		testHarness.processElement(new StreamRecord<>("Hello-0-0"), 0, 0);
+		testHarness.endInput(0,  0);
+		testHarness.processElement(new StreamRecord<>("Hello-0-1"), 0, 1);
+		testHarness.endInput(0,  1);
+
+		headOperator.waitOutputRecords(2);
+
+		testHarness.processElement(new StreamRecord<>("Hello-1-0"), 1, 0);
+		testHarness.processElement(new StreamRecord<>("Hello-1-1"), 1, 1);
```

Comment:
Wouldn't it be better to reorder this to:
```
		testHarness.processElement(new StreamRecord<>("Hello-1-0"), 1, 0);
		testHarness.endInput(1,  0);
		testHarness.processElement(new StreamRecord<>("Hello-1-1"), 1, 1);
		testHarness.endInput(1,  1);
```
? As it is now, we are not sure after which `endInput()` call operators received `endInput()` notification.